### ISSUE TITLE
Feature Suggestion: Per-Sticker Break and Conditional Stickers

### ISSUE THREAD

**Issue by Zulu-Inuoe** [joaotavora/sly#191]
I believe it would be very useful to have the ability to expand upon the current 'breaking stickers' by adding two features:

1. The ability to toggle breaking on/off on individual stickers rather than -all- stickers
2. The ability to conditionally 'trigger' a sticker

With the first, you have the ability to get a stack frame at a spot you're interested without manually inserting a `(break)` without having to hunt down every other sticker and removing it.
With the second you gain the ability to catch specific situations of interest without inserting your own `(when <condition> ..)`.

With the latter there is the issue of bad expressions for conditions. To make it user-friendly I think it should just fail to arm gracefully with a warning. To detect errors one idea that comes to mind is to take something like
```lisp
(defun foo (x)
  (/ pi x))
```
and when a condition of `(zerop x)` is set, verify the condition by doing:
```lisp
(cl:compile nil
  '(lambda (x)
    (when (zerop x)
      ...)
    (/ pi x)))
```
and verify `failure-p` and/or `warnings-p`

Though I'm not familiar with sly internals and am sure there's many smarter than me that can come up with better.

Thanks!

**Comment by djeis97** [comment]
How do conditional stickers interact with the pre- and post- breaking behavior? Does the condition run once, before any sticker breaking, and then that just determines if the sticker breaks on exit too?

On a related note, another option I think would be useful is some kind of print-on-sticker option akin to the tracing system- lighter weight than breaking, but helps in cases where the value is mutable and the recording later wouldn’t show the in-progress state.

**Comment by Zulu-Inuoe** [comment]
Yes, it would make sense to not invoke the sticker on exit if it failed to meet the condition to enter, if I'm understanding that correctly:
evaluate condition
if condition truthy, then run stickers related hooks (breaks, recording)
otherwise no sly hooks (beyond condition check of course)

I can't speak on the tracing bit. A common use-case I have for breakpoints is to freeze the program at a state so I can gather information about it (primarily in real-time type programs like a game loop). And that's what conditional stickers would get me.
Additionally I think conditional recording is useful for similar cases where you'll want to analyze a subset of executions meeting some criteria

**Comment by artforlife** [comment]
These are good ideas, especially the conditional stickers. If we could do conditional tracing, it would be a killer feature as well.

**Comment by Zulu-Inuoe** [comment]
I'm slowly grokking the sly sources. When I get a chance to concentrate on it I'll give it a whirl.
